packages/driver/src/dom/visibility.js

```diff
@@ -123,52 +119,196 @@ const elHasVisibilityHidden = ($el) => {
   return $el.css('visibility') === 'hidden'
 }
 
-const numberRegex = /-?[0-9]+(?:\.[0-9]+)?/g
-// This is a simplified version of backface culling.
+const elIsHiddenByTransform = ($el) => {
+  const list = extractTransformInfoFromElements($el)
+
+  if (existsInvisibleBackface(list)) {
+    return elIsBackface(list) ? 'backface' : 'visible'
+  }
+
+  return elIsTransformedToZero(list) ? 'transformed' : 'visible'
+}
+
+const extractTransformInfoFromElements = ($el, list = []) => {
+  list.push(extractTransformInfo($el))
+
+  const $parent = $el.parent()
+
+  if (!$parent.length || $document.isDocument($parent)) {
+    return list
+  }
+
+  return extractTransformInfoFromElements($parent, list)
+}
+
+const extractTransformInfo = ($el) => {
+  const el = $el[0]
+  const style = getComputedStyle(el)
+
+  return {
+    el,
+    backfaceVisibility: style.getPropertyValue('backface-visibility'),
+    transformStyle: style.getPropertyValue('transform-style'),
+    transform: style.getPropertyValue('transform'),
+  }
+}
+
+const existsInvisibleBackface = (list) => {
+  return !!_.find(list, { backfaceVisibility: 'hidden' })
+}
+
+const numberRegex = /-?[0-9]+(?:\.[0-9]+)?(?:[eE][+-]?[0-9]+)?/g
+const defaultNormal = [0, 0, 1]
+// This function uses a simplified version of backface culling.
 // https://en.wikipedia.org/wiki/Back-face_culling
 //
-// We defined view normal vector, (0, 0, -1), - eye to screen.
-// and default normal vector, (0, 0, 1)
+// We defined view vector, (0, 0, -1), - eye to screen.
+// and default normal vector of an element, (0, 0, 1)
 // When dot product of them are >= 0, item is visible.
-const elIsBackface = ($el) => {
-  const el = $el[0]
-  const style = getComputedStyle(el)
-  const backface = style.getPropertyValue('backface-visibility')
-  const backfaceInvisible = backface === 'hidden'
-  const transform = style.getPropertyValue('transform')
+const elIsBackface = (list) => {
+  const nextPreserve3d = (i) => {
+    return i + 1 < list.length &&
+    list[i + 1].transformStyle === 'preserve-3d'
+  }
 
-  if (!backfaceInvisible || !transform.startsWith('matrix3d')) {
-    return false
+  let i = 0
+
+  const finalNormal = (startIndex) => {
+    i = startIndex
+    let normal = findNormal(list[i].transform)
+
+    while (nextPreserve3d(i)) {
+      i++
+      normal = findNormal(list[i].transform, normal)
+    }
+
+    return normal
   }
 
-  const m3d = transform.substring(8).match(numberRegex)
-  const defaultNormal = [0, 0, -1]
-  const elNormal = findNormal(m3d)
+  const skipToNextFlat = () => {
```

Comment:
This file is starting to get really long. I think it would be good to extract some or most of these lower-level functions into a separate file or perhaps multiple files that are specific to transforms, backface, etc. Then we can keep this file's abstractions at a higher level and focused on the concept of visibility.